torchvision/tv_tensors/_dataset_wrapper.py

```diff
@@ -199,7 +200,16 @@ def __len__(self):
         return len(self._dataset)
 
     def __reduce__(self):
-        return wrap_dataset_for_transforms_v2, (self._dataset, self._target_keys)
+        # __reduce__ gets called when we try to pickle the dataset.
+        # In a DataLoader with spawn context, this gets called `num_workers` times from the main process.
+        # We have to reset the [target_]transform[s] attribute of the dataset
+        # before we pass it back to wrap_dataset_for_transforms_v2, because we
+        # set them to None in __init__().
+        dataset = copy(self._dataset)
+        dataset.transform = self.transform
+        dataset.transforms = self.transforms
+        dataset.target_transform = self.target_transform
+        return wrap_dataset_for_transforms_v2, (dataset, self._target_keys)
```

Comment:
TBH, I don't really understand why we needed to have `__reduce__` in the first place. I understand it's needed to support pickle, but my understanding stops here.

The whole logic seems really strange, i.e. calling yet again `wrap_dataset_for_transforms_v2()`, which is how the instance-being-pickled was created in the first place anyway. And now on top of that we add the "field resetting logic" i.e. we just undo what we did in `__init__`.

> I understand it's needed to support pickle, but my understanding stops here.

@pmeier can you remind me the details of this? Do you think we could support pickleability of these datasets in a different way that would require this "fix"?

Reply:
`pickle` does not support dynamic types like we do in

https://github.com/pytorch/vision/blob/3fb88b3ef1ee8107df74ca776cb57931fe3e9e1e/torchvision/tv_tensors/_dataset_wrapper.py#L104

by default. Thus, we implement the `__reduce__` method and tell it how to construct the object by its parts. Now only the items on L212 need to be pickled. While unpickling, the dynamic type is created anew and thus we circumvent the issue.

We have the dynamic type in the first place to support `isinstance` checks. There was also a different option in https://github.com/pytorch/vision/pull/7239#discussion_r1105093816 that would work without a dynamic type. I think that could potentially work without your fix, but I would need to test. However, this option also has its drawbacks (see discussion in the original PR).